@livestore/svelte

Rich-Harris opened the pull request:
Similar to #369 (cc @elliott-with-the-longest-name-on-github), this adds a `@livestore/svelte` package that provides a way for Svelte developers to use LiveStore in their apps. It takes a slightly different approach — rather than adding a new `store.createQuery` method, it essentially just monkey-patches `store.query` to make it reactive in a Svelte context. As such, there is no additional API surface area, and you can use it inline, like so:

```svelte
{#each store.query(visibleTodos$) as todo (todo.id)}
  <li>...</li>
{/each}
```

If `store.query(...)` is called in a non-reactive context (e.g. an event handler, or while initializing a component) it's as though you just called the underlying method. But if you call it in a reactive context (a `$derived`, or an expression in the template) then it will call `store.subscribe(...)` and cause the expression to re-run when the query value changes.

Another difference with #369 is that `createStore` can be called outside an effect. I'm in two minds about this — it's very convenient if you can do `export const store = await createStore(...)` from a module, but it does mean that the user is responsible for memory management if a store is created outside the component tree. It's possible that a more explicit distinction between those two modes of use would be preferable.

## TODO

- [ ] figure out the type errors
- [ ] add some tests
- [ ] decide whether or not we like this API

## Alternative API

One wrinkle is that `store.query(...)` can take a `QueryBuilder`, but `store.subscribe(...)` doesn't do anything with it. Which means that if you have something like this...

```svelte
<span class="todo-count">
  {store.query(tables.todos.count().where({ completed: false, deletedAt: null })} items left
</span>
```

...it will display the correct number (since it's just calling the underlying `query` method) but when it changes, nothing will happen. I don't know if that's an indication that `store.subscribe` _should_ be able to subscribe to a `QueryBuilder` (I don't totally understand the difference between it and `LiveQuery`/`LiveQueryDef`/`SignalDef` etc), or if this isn't actually the right API for Svelte, and we should instead do something more like #369, or perhaps like this:

```ts
const incomplete = new LiveQuery(
  store,
  tables.todos.count().where({ completed: false, deletedAt: null }),
  { label: 'incompleteCount' }
);
```

```svelte
<span class="todo-count">{incomplete.current} items left</span>
```

This is arguably more Svelte-idiomatic, and is more efficient insofar as we don't need to keep recreating the subscription on every change. (The reactivity trick in this PR is a little bit outré, if perfectly valid.) It is slightly more verbose for the developer though, and involves a new API.

Rich-Harris committed:
web-todomvc-svelte

Rich-Harris committed:
add @livestore/svelte

Rich-Harris committed:
tweaks

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/livestorejs/livestore?pullRequest=849) <br/>All committers have signed the CLA.

Rich-Harris committed:
unused

Rich-Harris committed:
tweak

schickling committed:
Enhance Svelte integration and update dependencies

- Added support for the @livestore/svelte package in package.json and pnpm-lock.yaml.
- Updated biome.jsonc to exclude Svelte files until Biome supports them.
- Refactored Svelte example project to use specific versioning for @livestore/svelte.
- Improved TypeScript configurations and added new types for CreateStoreOptionsPromise.
- Cleaned up imports and ensured consistent usage of TypeScript features across Svelte components.
- Introduced a new mod.ts file for better module exports in the Svelte package.

schickling committed:
Update @livestore/svelte package version and refine dependencies

- Changed @livestore/svelte version from 0.4.0-dev.17 to 0.0.1 in web-todomvc-svelte package.json.
- Updated svelte version to 5.43.14 for consistency.
- Adjusted dependencies in @livestore/svelte package.json, including updating jsdom and @sveltejs/vite-plugin-svelte versions.
- Restored scripts and files sections in @livestore/svelte package.json for better project structure.

schickling committed:
Refactor Svelte package structure and enhance testing capabilities

- Introduced a new `create-store.svelte.ts` file for improved store creation with Svelte reactivity.
- Added a `tsconfig.tests.json` for better TypeScript support in tests.
- Updated `vitest.config.ts` to streamline testing setup for the Svelte package.
- Removed outdated `vite.config.ts` to simplify configuration.
- Implemented comprehensive tests for the `createStore` functionality, ensuring proper reactivity and cleanup.
- Added utility functions for rendering tests and managing query states in Svelte components.

schickling committed:
Enhance testing for createStore in Svelte

- Added new tests to verify effect cleanup on reruns and subscription management during query swaps.
- Introduced a `renderRerunHarness` utility for testing effect reruns without query reference swaps.
- Updated existing tests to ensure proper handling of signal-driven queryables and cleanup behavior.
- Improved the Harness component to support external dependency updates for effect reruns.

schickling committed:
Remove publint dependency from @livestore/svelte package and update TypeScript versions in pnpm-lock.yaml

schickling committed:
Refactor @livestore/svelte package configuration and update dependencies

- Changed sideEffects in package.json to false for better tree-shaking.
- Removed unused dependencies including @sveltejs/package and svelte-check.
- Updated build and test scripts for improved clarity and functionality.
- Revised README.md to reflect new usage instructions and development practices.

schickling committed:
Add Svelte integration documentation and example

- Updated CHANGELOG.md to include Svelte framework guide and TodoMVC example.
- Modified example descriptions to reflect Svelte support in web adapter documentation.
- Added a new TodoMVC example using Svelte to the examples data.

schickling committed:
Merge branch 'dev' into livesvelte

schickling committed:
Update dependencies in web-todomvc-svelte package.json to version 0.4.0-dev.19 for @livestore packages and adjust @cloudflare/workers-types and vite versions for compatibility.

schickling committed:
Update pnpm-lock.yaml to reflect new versions for @cloudflare/workers-types and @cloudflare/vite-plugin, ensuring compatibility with the latest dependencies. Adjusted vite version specifier for improved package management.

schickling committed:
Enhance Vitest setup for Svelte testing

- Implemented a fetch shim to serve the compiled wasm file for wa-sqlite in jsdom, addressing local file fetching limitations.
- Added necessary imports and ensured compatibility with the testing environment.

arnolicious commented:
Just out of curiousity, this would be a client-only affair, correct?
In a non-spa sveltekit context, I assume it would be difficult to have livestore run both on the server and the client, or have it run on the server and somehow pass the reactive queries and events back and forth between client and server?

schickling commented:
Yes, for now the LiveStore/Svelte integration is only working on the client side. However, bringing SSR and server-ish support for LiveStore is certainly a topic we hope to explore soon as well.
